### ISSUE TITLE
v8-benchmarks failure in 2 tests

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#85]
Using the (outdated [1]) v8 version 7 benchmarks dir, we fail 2 tests on hardware:

```
Crypto: Error: Crypto operation failed
RayTrace: Error: Scene rendered incorrectly
```

All tests pass on mips64 simulator.

We need to fix the bug, so hardware passes, and also fix the simulator so that it works like the hardware does (and fails these tests)

[1] Of course, we should be looking at Octane now, this was just in-tree, and easy

**Comment by paul99** [comment]
Using mips64 branch, commit 34b1c912a, I just ran release mode mjsunit tests on sim and hardware.

```
[15:05|% 100|+ 4236|-  11]:  mjsunit  0515b-mjsr-sim.log
[57:17|% 100|+ 4225|-  22]   msjunit  0515a-mjsr-hw.log
```

The log files are on Dropbox, dir yy-pl-v8, (Dusan & Svetozar have been invited -- I will later create a better shared folder name, obviously this was a spot for me & Yuyin when we were the only 2 on project, no discrimination intended :)

The test-diffs.py script shows this output:

```
$ test-diffs.py 0515b-mjsr-sim.log 0515a-mjsr-hw.log 
8 failed in file 0515b-mjsr-sim.log, not in 0515a-mjsr-hw.log:
mjsunit/double-truncation
mjsunit/limit-locals
mjsunit/limit-locals
mjsunit/limit-locals
mjsunit/sparse-array-reverse
mjsunit/regress/regress-2185-2
mjsunit/regress/regress-2185-2
mjsunit/regress/regress-2185-2

19 failed in file 0515a-mjsr-hw.log, not in 0515b-mjsr-sim.log:
mjsunit/binary-op-newspace
mjsunit/binary-op-newspace
mjsunit/constant-folding-2
mjsunit/constant-folding-2
mjsunit/constant-folding-2
mjsunit/div-mod
mjsunit/div-mod
mjsunit/div-mod
mjsunit/math-pow
mjsunit/math-pow
mjsunit/math-round
mjsunit/mod
mjsunit/mod
mjsunit/mod
mjsunit/numops-fuzz-part1
mjsunit/sin-cos
mjsunit/unbox-double-arrays
mjsunit/unbox-double-arrays
mjsunit/unbox-double-arrays
```

All need to be looked at, but lets start with div-mod, mod,  and math-pow.

Alternatively, someone could start looking at crypto test directly.

Also, note to Yuyin: the limit locals seems to fail only on the sim! Interesting!

**Comment by paul99** [comment]
here is Octane r33:

```
$ ../out/mipsel64.release/d8 run.js
Richards: 964
DeltaBlue: 1490
Crypto: Error: Crypto operation failed
RayTrace: Error: Scene rendered incorrectly
EarleyBoyer: 1625
RegExp: 210
Splay: 351
SplayLatency: 1259
NavierStokes: 1088
PdfJS: 289
Mandreel: 1007
MandreelLatency: 464
Gameboy: Error: Buffer was of incorrect sample length.
CodeLoad: 882
Box2D: 954
zlib: 487
Typescript: 928
```

One more hard-fail in Gameboy, and it gets very slow at end. We are missing Dusan's zlib fix from bleeding_edge r19006. That should get back ported. 

Also, we took a major regression in Richard's compared to v8-v7. 

Primary request is to fix the breakages. 

Secondary is to backport r19006 (easy), and investigate Richards regression (maybe not easy)

**Comment by xwafish** [comment]
about "investigate Richards regression":
there are some mjsunit/testes run very slow in simulator, maybe 2 minutes, but  arm64 run faster, maybe we can start from find why they run so slow.

**Comment by paul99** [comment]
I'm not sure if this is what you mean, but I neglected to post all the scores for v8-benchmark version 7. Here they are. I only ran this and octane a couple times each, but it seems the big Richards score diff is real:

```
/benchmarks$ ../out/mipsel64.release/d8 run.js
Richards: 1583
DeltaBlue: 1676
Crypto: Error: Crypto operation failed
RayTrace: Error: Scene rendered incorrectly
EarleyBoyer: 1688
RegExp: 237
Splay: 265
NavierStokes: 1109
```

**Comment by paul99** [comment]
@yuyin, regarding your comment above about slow sim, I think that is real, but it is a different bug. There is a really old bug filed about this, let me update it.

**Comment by xwafish** [comment]
Fixed mjsunit/ div-mod, mod, math-pow and, math-floor. now they can pass on both sim and hardware.
@92076ad96451e7a55e5766419d4a7fe09cf92de0

Crypto and RayTrace also pass with the commit

**Comment by xwafish** [comment]
use 92076ad96451e7a55e5766419d4a7fe09cf92de0, only mjsunit/harmony/generators-iteration failed on hardware.

**Comment by paul99** [comment]
Very nice fix Yuyin! I verified the Octane tests all pass, and I'm running the other tests on my side now. I will commit this (to mips64 branch) later today.

**Comment by paul99** [comment]
I pushed Yuyin's fix 92076ad96451e7a55e5766419d4a7fe09cf92de0 to mips64 branch.

I've not tested the sun spider yet, but it does fix all v8-benchmarks and Octane tests.

Closing the issue.
